### PR TITLE
Updated CA1010 to limit diagnostics to the direct implementors of non…

### LINKED ISSUE #1325: CA1010 false positive
#### Analyzer package

Microsoft.CodeAnalysis.FxCopAnalyzers

#### Analyzer

CA1010

#### Repro steps

```C#
// external code, out of my control
public class ThirdPartyType : IEnumerable
{
    ...
}

public class MyClass : ThirdPartyType
{
}
```

#### Expected behavior

No violation, since it is not my code violating the rule.

#### Actual behavior

```
error CA1010: Collections should implement generic interface
```

### PR DESCRIPTION
**Pull request by 333fred**
…generic collections, to prevent cascading errors and reduce noise.
Fixes #1325.

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx [hunk 1]
```diff
@@ -514,7 +514,7 @@
     <value>To broaden the usability of a collection, implement one of the generic collection interfaces. Then the collection can be used to populate generic collection types.</value>
   </data>
   <data name="CollectionsShouldImplementGenericInterfaceMessage" xml:space="preserve">
-    <value>Collections should implement generic interface</value>
+    <value>Collection '{0}' inherits '{1}' without implementing '{2}'. Publicly-visible collections should implement the generic version to broaden usability.</value>
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx]: FXCop has a different message, but this seem more clearer to me. This is the FXCop message for reference: http://index/?query=CollectionsShouldImplementGenericInterface&rightProject=DesignRules&file=CollectionsShouldImplementGenericInterface.cs&line=100

Should the message be: `Collection '{0}' directly or indirectly implements '{1}' without implementing '{2}'`?

**Comment by 333fred** [comment]
Closing this PR in favor of new implementation.

File: src/Utilities/WellKnownTypes.cs [hunk 2]
```diff
@@ -46,6 +46,16 @@ public static INamedTypeSymbol GenericIList(Compilation compilation)
             return compilation.GetTypeByMetadataName("System.Collections.Generic.IList`1");
         }
 
+        public static INamedTypeSymbol CollectionBase(Compilation compilation)
```
Comment (sharwell) [inline src/Utilities/WellKnownTypes.cs]: ❓ Are these used?

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs [hunk 3]
```diff
@@ -96,18 +98,94 @@ public override void Initialize(AnalysisContext analysisContext)
                 return;
             }
 
-            var allInterfaces = namedTypeSymbol.AllInterfaces.Select(t => t.OriginalDefinition).ToImmutableHashSet();
+            var allInterfaces = namedTypeSymbol.AllInterfaces.Select(t => t.OriginalDefinition).ToImmutableArray();
```
Comment (sharwell) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]: :question: This changes the behavior from before, specifically in the case of types that implement both `IEnumerable<A>` and `IEnumerable<B>`. Was this change intentional? (I believe it does not change behavior but I still wanted to ask.)

:bulb: Consider using `SelectAsArray` for efficiency. Even better, simply `foreach` over `AllIterfaces` or the result of `Select` directly .

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs [hunk 4]
```diff
@@ -96,18 +98,94 @@ public override void Initialize(AnalysisContext analysisContext)
...
+                                                       namedTypeSymbol.Locations.First(),
+                                                       namedTypeSymbol.Name,
+                                                       implementedInterface.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat),
+                                                       missingInterface.ToDisplayString(SymbolDisplayFormat.MinimallyQualifiedFormat)));
+        }
+
+        private struct CollectionsInterfaceStatus
+        {
+            public bool IListPresent { get; set; }
+            public bool GListPresent { get; set; }
```
Comment (sharwell) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]: :bulb: `GenericIListPresent` (same for related)

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx [hunk 5]
```diff
@@ -514,7 +514,7 @@
     <value>To broaden the usability of a collection, implement one of the generic collection interfaces. Then the collection can be used to populate generic collection types.</value>
   </data>
   <data name="CollectionsShouldImplementGenericInterfaceMessage" xml:space="preserve">
-    <value>Collections should implement generic interface</value>
+    <value>Collection '{0}' directly or indirectly inherits '{1}' without implementing '{2}'. Publicly-visible collections should implement the generic version to broaden usability.</value>
```
Comment (sharwell) [inline src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx]: :bulb:

> Type '{0}' implements '{1}' without implementing '{2}'. Publicly-visible collection types should implement generic collection interfaces to improve usability.

Comment (333fred) [reply src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx]: That's what I originally had it as. We decided to include the reference about indirect inheritance to remind people that that's a possibility.

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs (hunk 3 above)
Comment (333fred) [reply src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/CollectionsShouldImplementGenericInterface.cs]: The change was not intentional, but I agree that it should not make a difference in implementation, except in cases where there are lots of implementations of any of the generic interfaces, but I don't think that will make much a difference here.

**Comment by 333fred** [comment]
@sharwell addressed feedback.

File: src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx (hunk 5 above)
Comment (sharwell) [reply src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/MicrosoftApiDesignGuidelinesAnalyzersResources.resx]: Note that I made small changes to several sections of the text. The current form seems too wordy; you can (and should) place additional information in the document pointed to by the Help Link URI.
